### ISSUE TITLE
getAll queries result the same document multiple times for some queries

### ISSUE THREAD

**Issue by eliaslevy** [rethinkdb/rethinkdb#4147]
If you have a record in which you index two fields with a method into a single multi index, and both fields have the same value, when you search for the field value, the ```getAll``` will return duplicates for the matching document.  E.g. if you have doc like ```{ "a": 1, "b": 1}```, and you create a single multi index that indexes both "a" and "b", then search for 1, you get back two copies of the doc.

Similarly, if the doc is ```{ "a": 1, "b": 2}```, ```.getAll(1, 2)``` will return the document twice.

While there may be situations where this is desirable, I'd argue it fails the principle of least astonishment.  If only a single document matches, I except to receive only a single document in return.

**Comment by danielmewes** [comment]
Hi @eliaslevy , this is related to https://github.com/rethinkdb/rethinkdb/issues/3462 .

If you only get a few documents from the `getAll`, you can work around it by appending `.distinct()` after the `getAll`.

I feel like we had discussed this at some point, but I don't remember what the conclusion was.
@mlucy do you remember any discussion on this?

It seems like removing duplicates where a single document is indexed under the same key multiple times could be done efficiently by doing an ordered index traversal.
It might be inconsistent with the behavior we get for the case where a document is indexed under multiple *different* keys, since we would still get duplicates in that case. So I'm not sure if it would be less confusing or more confusing to filter same-key duplicates.

**Comment by mlucy** [comment]
I can't find the discussion, but I believe we talked about this before and decided to keep the current behavior since you can use `distinct` to remove duplicates but if we removed duplicates there's no way for the user to add them back in.  That said, I'd be open to discussing it again.

**Comment by eliaslevy** [comment]
@danielmewes aye, this seems to be the same root cause as #3462,

**Comment by thomasmodeneis** [comment]
I'm having the same issue here, using distinct will indeed solve the problem, but I'm not sure if this will affect performance of my cluster, any ideas ?

**Comment by danielmewes** [comment]
@thomasmodeneis Whether the `distinct` can become a performance problem depends mostly on the size of the `getAll` result set. How many documents are we talking about approximately?

**Comment by thelinuxlich** [comment]
I agree that having to run .distinct() after getAll() with multi index seems like a hack(and a inneficient one, since distinct is constrained to the arrayLimit parameter)

**Comment by thomasmodeneis** [comment]
I have a feedback table with around 3GB. For me its quite strange that some queries still return duplicated values as I'm manually validating for duplicates at insert time.

Maybe the main point here could be that I'm doing the query with skip and limit ? would this affect the general behavior of the database ?

**Comment by danielmewes** [comment]
@thomasmodeneis If you look at one of the documents that's duplicated in your queries, does it have the same multi index key multiple times? (e.g. `["a", "b", "a"]` where `"a"` occurs twice)
Or are you passing multiple arguments into `getAll(key1, key2, ...)` and the same document matches multiple different keys?

The former could be avoided more easily at insert/update time, by including the `.distinct()` into the insert query (I can give you some more details and an example if this is actually what's happening).

**Comment by danielmewes** [comment]
@Tryneus and I came up with an algorithm that would allow us to efficiently avoid duplicate results in a `getAll`:
1. On any given shard, we keep track of the previously traversed `[sindex key, primary key]` pair. We skip over any entry that has the same pair as the previous value. This gives us the invariant that for a given sindex key, a given shard will only give us distinct results.
2. During unsharding, make sure that we maintain the invariant that the output is always ordered by primary key. This again allows us to easily omit duplicates in the process. For a single sindex key, we already get results that way from each shard, and we just need to merge them across shards in an ordered way. For multi-gkey `getAll`s, see the next point:
3. If we have multiple sindex keys (for variadic `getAll`s), we must make sure to perform a "merge sort" like merge not only during unsharding, but also on each shard to maintain the ordering by primary key. This will require a change to how we're traversing the tree for multiple keys at the moment. We currently send all batches for the first key first. Then we transition to the second key and so on. In order to make the merge work, we'd need to traverse for each key in parallel before we can send a batch back to the parsing node.

This is still going to be a bit of work, but I believe it's feasible.

**Comment by mlucy** [comment]
Ideally we'd want to support omitting duplicates from multi-indexes for `between` as well as for `get_all`.

Here's an alternative algorithm that I think works for both:

* When we store a row with a multi-index `m`, we also store a sorted array of all the values in `m` plus their offsets in `m` `(so ["a", "c", "b", "c"]` becomes `[["a", 0], ["b", 2], ["c", 1], ["c", 3]]`.
* When you're in the middle a multi-index traversal over the range `[A, B)` and you see a value `X`, you look up `X` in that sorted array via binary search.  (We serialize the offset in the multi-index array into the key alongside the index value, so you know enough to distinguish the `c` at index 1 and the `c` at index 3 during traversal.)
* We look at the value immediately before `X` in that array.  If that value is in `[A, B)`, we drop `X` from the traversal.  If it isn't (or if `X` is the first element of the array) we include `X` in the traversal.

If we're OK doing work proportional to the size of the array, we can skip the first bullet point and just recompute the sorted array when we load the row off of disk.  For most multi-indexes (i.e. ones without thousands of entries) that should be plenty fast.

**Comment by thelinuxlich** [comment]
For inspiration, here's how ElasticSearch is handling this:

https://www.elastic.co/guide/en/elasticsearch/guide/current/complex-core-fields.html#object-arrays

https://www.elastic.co/guide/en/elasticsearch/guide/current/parent-child.html

**Comment by danielmewes** [comment]
> If we're OK doing work proportional to the size of the array, we can skip the first bullet point and just recompute the sorted array when we load the row off of disk. For most multi-indexes (i.e. ones without thousands of entries) that should be plenty fast.

I think for many queries we do that anyway, and we used to do it always until recently. I think that would be ok, and would avoid a lot of extra work.

Your proposal sounds good and reasonably easy to implement. I like it.

**Comment by jazoom** [comment]
I just came across a similar issue to this and I have to say, the behaviour is quite unexpected. Here's my issue:

I have a secondary index where it is possible for some items to have the same value, but they will always have another value that is different. (e.g. { secondaryKey: a, otherValue: n} and {  secondaryKey: a, otherValue: o }. If the secondary key is the same, otherValue is guaranteed to be different (according to my api).

I am doing this: getAll(r.args(secondaryKeys), { index: secondaryKey}). It is possible the secondaryKeys array there contains duplicates. The behaviour I expected was that one copy of every matching document would be returned. I was surprised that a copy was returned for every key, even if a duplicate. This should really be in your docs. It would have saved quite a bit of bug-hunting. Based on the above conversation, I can see you already know about this.

Furthermore, and more surprisingly, the result of the above query DOESN'T return the other matching document (i.e.  {  secondaryKey: a, otherValue: o } ).

So we end up with 2 of this document:

{ secondaryKey: a, otherValue: n}

and none of this document:

 {  secondaryKey: a, otherValue: o }

If this is the way getAll is intended to work then you might consider changing the name? That behaviour is not what I think of when I read "getAll".

RethinkDB is awesome. Keep up the great work, team!

**Comment by danielmewes** [comment]
@jazoom Thanks for the feedback. Good idea documenting the fact that duplicates in the key list result in duplicate results better. I've opened https://github.com/rethinkdb/docs/issues/1067 for this.

If you haven't figured this out already, you might want to look into `distinct` to remove duplicates in your key list: `tbl.getAll(r.args(r.expr(keyList).distinct()), {index: ...})`.

About your second observation regarding getting two copies of the first and no copy of the second document: That definitely sounds wrong. I can't reproduce this unfortunately.
Here's what I'm seeing:
```js
r.tableCreate('t1');
r.table('t1').indexCreate("a");
r.table('t1').insert({a: "a", o: 1});
r.table('t1').insert({a: "a", o: 2});

r.table('t1').getAll("a", "a", {index: "a"});
// Returns:
{
    "a": "a" ,
    "id": "02e9d4f7-41fc-4501-80af-e10b3eb44760" ,
    "o": 1
}
{
    "a": "a" ,
    "id": "717ed689-c07a-4796-bcc0-8711804bfc7a" ,
    "o": 2
}
{
    "a": "a" ,
    "id": "02e9d4f7-41fc-4501-80af-e10b3eb44760" ,
    "o": 1
}
{
    "a": "a" ,
    "id": "717ed689-c07a-4796-bcc0-8711804bfc7a" ,
    "o": 2
} 
```

Does anything stand out to you that's different in your case?

**Comment by thelinuxlich** [comment]
Probably it is a multi index

**Comment by jazoom** [comment]
Hmmm... it's just a basic secondary index. I'll play with it some more to find a difference, or, you know, a mistake in my code. I'll let you know.

**Comment by jazoom** [comment]
My apologies, @danielmewes. It appears to work as you said. I'm not sure what was going on before.

Thanks for adding info on duplicate behaviour to getAll in your documentation. I think that will save other people some hassle in future.
